### PR TITLE
chore: Add integration tests for Open Telemetry

### LINKED ISSUE #38: Add integration test for Open Telemetry
**Describe the solution you'd like**
Open telemetry works always.

**Additional context**
We recently had an uncaught regression as there were no integration tests for Otel.
We should take the time to add a test for each type of collector.

### PR DESCRIPTION
**Pull request by o0Ignition0o**
will fix #38 eventually

### CODE REVIEW

**Comment by o0Ignition0o** [comment]
Brace yourselves, it's macro time.

File: apollo-router-core/src/query_planner/router_bridge_query_planner.rs [hunk 1]
```diff
@@ -23,7 +23,12 @@ impl RouterBridgeQueryPlanner {
 
 #[async_trait]
 impl QueryPlanner for RouterBridgeQueryPlanner {
-    #[tracing::instrument(name = "plan", level = "debug")]
+    #[tracing::instrument(
+        skip_all,
+        fields(query = %query),
```
Comment (Geal) [inline apollo-router-core/src/query_planner/router_bridge_query_planner.rs:28]: I think there was a decision a few weeks ago about not sending the query as part of the traces if we do not have a good way to sanitize them (ex: remove variables with potentially private information). I cannot find the original discussion but there's this issue: https://github.com/apollographql/router/issues/48

Comment (o0Ignition0o) [reply apollo-router-core/src/query_planner/router_bridge_query_planner.rs:28]: Oh that makes sense, lemme remove that

File: apollo-router-core/src/spec/query.rs [hunk 2]
```diff
@@ -26,7 +26,13 @@ impl Query {
     ///
     /// This will discard unrequested fields and re-order the output to match the order of the
     /// query.
-    #[tracing::instrument(skip_all, level = "trace")]
+    #[tracing::instrument(
+        skip_all,
+        fields(
+            errors = %response.errors.iter().map(std::string::ToString::to_string).collect::<Vec<_>>().join(", ")
```
Comment (o0Ignition0o) [inline apollo-router-core/src/spec/query.rs:32]: @Geal is this one worth it ?

Comment (Geal) [reply apollo-router-core/src/spec/query.rs:32]: I do not think so, and anyway that would not be the right place to trace those errors, right?

File: apollo-router/tests/snapshots/integration_tests__traced_basic_request.snap [hunk 3]
```diff
@@ -0,0 +1,961 @@
...
+              [
+                "otel.kind",
+                {
+                  "Value": "client"
+                }
+              ],
+              [
+                "otel.name",
+                {
+                  "Debug": "POST /graphql"
```
[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router/tests/snapshots/integration_tests__traced_basic_request.snap:895]: https://docs.rs/reqwest-tracing/latest/src/reqwest_tracing/middleware.rs.html#23, not really sure tbh

[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router/tests/snapshots/integration_tests__traced_basic_request.snap:895]: oook i see, it s the kind of visitor entry

**Review by Geal — APPROVED** [review]

**Comment by o0Ignition0o** [comment]
missing _a lot_ of docs and explanation, but the functionality is there

File: Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,11 @@
 [workspace]
 default-members = ["apollo-router", "apollo-router-core"]
-members = ["apollo-router", "apollo-router-core", "apollo-router-benchmarks", "xtask"]
+members = [
+    "apollo-router",
+    "apollo-router-core",
+    "apollo-router-benchmarks",
+    "xtask",
+]
```
Comment (cecton) [inline Cargo.toml:8]: 1. :yarn: You keep changing that and someone else keep changing it back.
2. :screwdriver: This file is unnecessarily touched in this PR.

File: apollo-router-core/src/query_planner/mod.rs [hunk 5]
```diff
@@ -323,14 +320,16 @@ mod fetch {
...
 
-            let Variables { variables, paths } = Variables::new(
-                &self.requires,
-                self.variable_usages.as_ref(),
-                data,
-                current_dir,
-                request,
-                schema,
-            )?;
+            let Variables { variables, paths } = query_span.in_scope(|| {
```
Comment (cecton) [inline apollo-router-core/src/query_planner/mod.rs:323]: :yarn: per our ADR on telemetry, we should prefer to put this at the source instead of the caller.

File: apollo-router-core/src/query_planner/mod.rs [hunk 6]
```diff
@@ -343,16 +342,18 @@ mod fetch {
...
-                return Err(FetchError::SubrequestUnexpectedPatchResponse {
-                    service: service_name.to_owned(),
-                });
-            }
+            query_span.in_scope(|| {
+                if !response.is_primary() {
+                    return Err(FetchError::SubrequestUnexpectedPatchResponse {
+                        service: service_name.to_owned(),
+                    });
+                }
```
Comment (cecton) [inline apollo-router-core/src/query_planner/mod.rs:353]: :yarn: I'm don't understand why this is under the query_span

File: apollo-router-core/src/query_planner/router_bridge_query_planner.rs [hunk 7]
```diff
@@ -23,7 +23,7 @@ impl RouterBridgeQueryPlanner {
 
 #[async_trait]
 impl QueryPlanner for RouterBridgeQueryPlanner {
-    #[tracing::instrument(name = "plan", level = "debug", skip_all)]
+    #[tracing::instrument(skip_all, name = "plan", level = "debug")]
```
Comment (cecton) [inline apollo-router-core/src/query_planner/router_bridge_query_planner.rs:26]: :yarn: File unnecessarily touch in this PR

If possible I would like to keep the commits on main focus on their actual feature. If you want to re-order all the `skip_all` everywhere, you can make it on a separate PR.

File: apollo-router/tests/integration_tests.rs [hunk 8]
```diff
@@ -43,7 +43,7 @@ macro_rules! assert_federated_response {
     };
 }
 
-#[test(tokio::test)]
+#[tokio::test]
```
Comment (cecton) [inline apollo-router/tests/integration_tests.rs:46]: :grey_question: Why did you remove the logger initialization for these tests?

**Review by cecton — COMMENTED** [review]
Can we take more time for PR reviews?

I understand this one has been created 28 days ago but it is ready for review for 3 days only. This doesn't give enough time for the reviewers.
